packages/core-components/src/components/Link/Link.tsx

```diff
@@ -75,17 +79,37 @@ const getNodeText = (node: React.ReactNode): string => {
   return '';
 };
 
-/**
- * Thin wrapper on top of material-ui's Link component, which...
- * - Makes the Link use react-router
- * - Captures Link clicks as analytics events.
- */
-export const Link = React.forwardRef<any, LinkProps>(
-  ({ onClick, noTrack, ...props }, ref) => {
+export interface LinkContext {
+  to: string;
+  children: React.ReactNode;
+}
```

Comment:
Wanna dig into this more as it was part of what I wanted to get at with the initial discussion. What are the benefits of this extra layer? More specifically the context type and the `value` prop. Could we instead use the existing props directly and only?

Reply:
The idea is that this extra layer is what the component _exports_ as actually changeable.

 * Props are what the user uses to the create component at _use-time_.
 * The context values are what _extensions_ use to _alter_ the component. It's kind of the internal _state_.

Consider `<UserCard userEntityRef={myRef} />` where the only prop here is `userEntityRef`, which you probably don't want to change (it wouldn't be _that_ useful). What this PR provides, is a way to change how that UserCard actually works, what it shows (e.g. action buttons, titles, avatars), etc.

So the UserCard parses the `userEntityRef`, get's the user info from the catalog, prepares data structures and then creates internal state out of this. Most components do this already, but they do it with `useState`. Then that _state_ is what is used to _render_ the output. If we instead formalise this state into one single context object and push that to the second (inner) _rendering component_, that's when we can let other intermediate components _alter_ the state and therefore change what the component does.

So, you can argue that this adds complexity or at least adds another layer, but you can also reason that it's actually just formalising the internal state components already have, into a _"publicly"_ interceptable state. You likely won't add complexity, just refactor and extract the preparation phase and rendering phase into two separate components.

And all parts of this;
 * the "setup" phase (called Provider here, maybe Setup is better)
 * the "render" phase (called Component here, maybe Render is better)
 * all extensions

all of those are pure React components with proper React lifecycle handling, so they can use hooks, asynchrony, etc.

In the Link example, it makes sense to let extensions change the `children` so they can decorate the UI, but in most components, this probably won't make much sense.

Reply:
I don't actually see how the context helps with your `userEntityRef` example, seems you could do similar things using just a middleware-like approach?

The one thing I do see is the potential information loss, for example if the context has `{x: number, y: number}` but the props is just `{sum: number}`, then you can't really reverse that operation. That might just be a bit of a straw man though tbh, because you could pass `x`, `y` through props. 

There's also the versioning and API stability aspect, but tbh I don't see a big benefit here either. The props are obviously already public API, and unless we remove the possibility to read props or override the render component then they're also both input and output types. Adding context on top of that just means we have a second type to manage, again both input and output. Even if that type is simpler, it's still extra.

so TL;DR: to me it seems like the context addition is only worthwhile if we keep the component props write-only, in which case it arguably simplifies the API surface.